drivers/gpio/gpio_eos_s3.c

```diff
@@ -0,0 +1,335 @@
+{
+	ARG_UNUSED(dev);
+	uint32_t target_value;
+	uint32_t output_states = MISC_CTRL->IO_OUTPUT;
+
+	target_value = output_states ^ value;
+	for (uint8_t gpio_num = 0; gpio_num < MAX_GPIOS; gpio_num++) {
+		if (target_value & 1) {
+			HAL_GPIO_Write(gpio_num, 1);
+		} else {
+			HAL_GPIO_Write(gpio_num, 0);
+		}
+		target_value = target_value >> 1;
+	}
+
+	return 0;
+}
+
+static int gpio_eos_s3_manage_callback(const struct device *dev,
+				       struct gpio_callback *callback, bool set)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+
+	return gpio_manage_callback(&data->callbacks, callback, set);
+}
+
+static int gpio_eos_s3_pin_interrupt_configure(const struct device *dev,
+					       gpio_pin_t gpio_num,
+					       enum gpio_int_mode mode,
+					       enum gpio_int_trig trig)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+	const struct gpio_eos_s3_config *config = dev->config;
+	uint8_t pin_cfg = config->pin_cfgs[gpio_num];
+	GPIOCfgTypeDef gpio_cfg;
+	PadConfig pad_conf;
+
+	/* pin_cfg should be either 0 (primary) or 1 (secondary) */
+	if (pin_cfg > 1) {
+		return -EINVAL;
+	}
+
+	/* Check for an invalid gpio number */
+	if (gpio_num >= MAX_GPIOS) {
+		return -EINVAL;
+	}
+
+	gpio_cfg.ucGpioNum = gpio_num;
+	pad_conf = pad_configs[(8 * pin_cfg) + gpio_num];
+	gpio_cfg.xPadConf = &pad_conf;
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		/* Reset IRQ configuration */
+		int irq_num = HAL_GPIO_IntrCfg(&gpio_cfg);
+
+		if (irq_num < 0) {
+			return -EINVAL;
+		}
+
+		data->gpio_irq_num[gpio_cfg.ucGpioNum] = BIT((uint8_t)irq_num);
+		/* Disable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 &= ~data->gpio_irq_num[gpio_num];
+	} else {
+		/* Prepare configuration */
+		if (mode == GPIO_INT_MODE_LEVEL) {
+			gpio_cfg.intr_type = LEVEL_TRIGGERED;
+			if (trig == GPIO_INT_TRIG_LOW) {
+				gpio_cfg.pol_type = FALL_LOW;
+			} else {
+				gpio_cfg.pol_type = RISE_HIGH;
+			}
+		} else {
+			gpio_cfg.intr_type = EDGE_TRIGGERED;
+			switch (trig) {
+			case GPIO_INT_TRIG_LOW:
+				gpio_cfg.pol_type = FALL_LOW;
+				break;
+			case GPIO_INT_TRIG_HIGH:
+				gpio_cfg.pol_type = RISE_HIGH;
+				break;
+			case GPIO_INT_TRIG_BOTH:
+				return -ENOTSUP;
+			}
+		}
+
+		/* Set IRQ configuration */
+		int irq_num = HAL_GPIO_IntrCfg(&gpio_cfg);
+
+		if (irq_num < 0) {
+			return -EINVAL;
+		}
+
+		data->gpio_irq_num[gpio_cfg.ucGpioNum] = BIT((uint8_t)irq_num);
+
+		/* Enable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 |= data->gpio_irq_num[gpio_num];
+	}
+
+	return 0;
+}
+
+static void gpio_eos_s3_isr(const struct device *dev)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+	uint8_t intr_status = INTR_CTRL->GPIO_INTR;
+
+	for (uint8_t i = 0; i < MAX_GPIOS; i++) {
+		uint8_t intr_num = BIT(i);
+
+		if (intr_status & intr_num) {
+			for (int gpio_num = 0; gpio_num < MAX_GPIOS; gpio_num++) {
+				if (intr_num == data->gpio_irq_num[gpio_num]) {
+					gpio_fire_callbacks(&data->callbacks,
+							    dev, BIT(gpio_num));
+				}
+			}
+			INTR_CTRL->GPIO_INTR |= intr_num;
```

Comment:
I assume we are writing to the register to clear the interrupt. This should be done immediately after the line
```
uint8_t intr_status = INTR_CTRL->GPIO_INTR;
```
before we call `gpio_fire_callbacks` to avoid losing interrupts. We need to assume the `gpio_fire_callbacks` function can take long to execute.